learningequality/studio
# Allow users to retry copy failures

## Summary

This PR brings in the capability to retry failed copies and detection of failing copies.

### Manual verification steps performed

To generate copy error state on the frontend, I raised an exception from the backend copy function. These are the checks I've performed:-
 
1. Checked whether error in the `changesForSyncing` table are detected.
3. Checked whether importing from channels errors out.
4. Checked whether copying completes.
5. Checked whether retrying completes successfully when the exception is removed from the backend code.
6. Checked whether after multiple failures, retrying still reaches success upon exception removal.
7. Copying multiple nodes together by selecting them works as expected.
8. When copy is failed, the node can be removed.

### Screenshots (if applicable)

The below screenshot is the current design, it's matching with the expected design. (thanks to sir @bjester for helping me out fix the design).

![image](https://github.com/learningequality/studio/assets/26724128/ddbc634e-dece-4a9a-b5e3-ccdbc973cb80)

The below is the UI design that is expected.

![image](https://user-images.githubusercontent.com/6668144/110513952-a2975b00-80bb-11eb-9dda-4c078facce96.png)

### Does this introduce any tech-debt items?

Possible tech-debts:-

- the CSS that is used to prevent wrapping `retry` action can be refactored to make it responsive and robust.
- due to the current architecture, copy error state doesn't persist across sessions. 

___
## Reviewer guidance

Does manual verification steps checks out green? 


## References

Closes https://github.com/learningequality/studio/issues/2850.
Closes https://github.com/learningequality/studio/issues/3914.

----

### Contributor's Checklist


PR process:


Studio-specifc:



Testing:

___

### Reviewer's Checklist
#### This section is for reviewers to fill out.
